File: ioi/2016/day2/messy/solutions-to-submit/messy_c_ok.c.dir/messy.c
// name = messy_c_ok.c, type = c.gcc

#include "messy_c.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>


int nn;
char* address;

void helper(int length, int w) {
	memset(address, '1', nn);
	for (int i = w; i < nn; i += (1 << length)) {
		address[i] = '0';
	}
	int j = w;
	for (int i = 0; i < (nn / (1 << length)); i++) {
		address[j] = '1';
		if (i % 2 == 1) {
			add_element(address);
		}
		address[j] = '0';
		j += 1 << length;
	}
}

void doWrites(int n_) {
	nn = n_;
	address = malloc(nn + 1);
	address[nn] = 0;
	for (int i = 0; i < nn; i++) {
		memset(address, '0', nn);
		address[i] = '1';
		if (i % 2 == 1) {
			add_element(address);
		}
	}
	int x = 1;
	int log = 0;
	while (x < nn) {
		x *= 2;
		log++;
	}
	for (int length = 1; length < log; length++) {
		for (int i = 0; i < (1 << length); i++) {
			helper(length, i);
		}
	}
}

void read(int* t, int tSize, int* answer, int step, int w) {
	if (tSize == 1) {
		answer[t[0]] = w;
		return;
	}
	int* t0 = malloc(tSize / 2 * sizeof(int));
	int* t1 = malloc(tSize / 2 * sizeof(int));
	int at0 = 0;
	int at1 = 0;
	memset(address, '1', nn);
	for (int j = 0; j < tSize; j++) {
		address[t[j]] = '0';
	}
	for (int i = 0; i < tSize; i++) {
		int j = t[i];
		address[j] = '1';
		if (!check_element(address)) {
			t0[at0++] = j;
		}
		else {
			t1[at1++] = j;
		}
		address[j] = '0';
	}
	read(t0, tSize / 2, answer, step + 1, w);
	read(t1, tSize / 2, answer, step + 1, w + (1 << step));
}

void doReads(int n, int* answer) {
	int* order = malloc(n * sizeof(int));
	for (int i = 0; i < n; i++) {
		order[i] = i;
	}
	read(order, n, answer, 0, 0);
}

void restore_permutation(int n, int w, int r, int* answer) {
	doWrites(n);
	compile_set();
	doReads(n, answer);
}

File: ioi/2016/day2/messy/solutions-to-submit/messy_pas_ok.pas.dir/messy.pas
// name = messy_pas_ok.pas, type = pas.fpc

unit messy;

interface

type TIntArray = array of longint;

function restore_permutation(n, w, r: longint) : TIntArray;


implementation

uses graderlib;

type
  nodep = ^node;
  node = Record
    zero, one: ^node;
  end;


var
  n, w, r, i: longint;
  p, answer: array of longint;
  ok: boolean;
  nn: longint;
  address: string;
  initialized: boolean = false;
  compiled: boolean;
  nodeSet: nodep;


  function create(): nodep;
var
  res: nodep;
begin
  res := new(nodep);
  res^.zero := NIL;
  res^.one := NIL;
  create := res;
end;

procedure wa();
begin
  writeln('WA');
  halt;
end;



procedure fill(c: char);
var
  i: longint;
begin
  for i := 1 to nn do
    address[i] := c;
end;

procedure helper(length, w: longint);
var
  i, j, shift: longint;
begin
  fill('1');
  i := w;
  shift := 1 shl length;
  while i < nn do begin
    address[i + 1] := '0';
    inc(i, shift);
  end;
  j := w;
  for i := 0 to nn div shift - 1 do begin
    address[j + 1] := '1';
    if i mod 2 = 1 then
       add_element(address);
    address[j + 1] := '0';
    inc(j, shift);
  end;
end;

procedure doWrites(n_: longint);
var
  i, x, log, length, shift: longint;
begin
  nn := n_;
  setLength(address, nn);
  for i := 0 to nn - 1 do begin
    fill('0');
    address[i + 1] := '1';
    if i mod 2 = 1 then
       add_element(address);
  end;
  x := 1;
  log := 0;
  while x < nn do begin
    x := x * 2;
    inc(log);
  end;
  for length := 1 to log - 1 do begin
    shift := 1 shl length;
    for i := 0 to shift - 1 do
      helper(length, i);
  end;
end;

procedure readImpl(t: array of longint; tSize: longint; var answer: array of longint; step, w: longint);
var
  t0, t1: array of longint;
  at0, at1, i, j: longint;
begin
  if tSize = 1 then begin
    answer[t[0]] := w;
    exit;
  end;
  setLength(t0, tSize div 2);
  setLength(t1, tSize div 2);
  at0 := 0;
  at1 := 0;
  fill('1');
  for j := 0 to tSize - 1 do
    address[t[j] + 1] := '0';
  for i := 0 to tSize - 1 do begin
    j := t[i];
    address[j + 1] := '1';
    if not check_element(address) then begin
      t0[at0] := j;
      inc(at0);
    end else begin
      t1[at1] := j;
      inc(at1);
    end;
    address[j + 1] := '0';
  end;
  readImpl(t0, tSize div 2, answer, step + 1, w);
  readImpl(t1, tSize div 2, answer, step + 1, w + (1 shl step));
end;

procedure doReads(n: longint; var answer: array of longint);
var
  order: array of longint;
  i: longint;
begin
  setLength(order, n);
  for i := 0 to n - 1 do
    order[i] := i;
  readImpl(order, n, answer, 0, 0);
end;

function restore_permutation(n, w, r: longint) : TIntArray;
var
    answer : array of longint;
begin
  doWrites(n);
  compile_set;
  setlength(answer, n);
  doReads(n, answer);
  restore_permutation := answer;
end;


end.

